<?xml version="1.0"?>
<launch>
    <!-- Load joint controller configurations from YAML file to parameter server -->
    <rosparam file="$(find diff_drive)/config/controllers.yaml" command="load"/>
    <include file="$(find diff_drive)/launch/view_robot.launch" />
    <rosparam param="wheel_radius">0.035</rosparam>
    <rosparam param="wheel_separation">0.15</rosparam>
    <rosparam param="ticks_meter">3665</rosparam>
    <!--<node pkg="rosserial_python" type="serial_node.py" name="serial_node" launch-prefix="xterm -e"/> 
  -->
    <!--
  <node pkg="rosserial_python" type="serial_node.py" name="serial_node">
		<param name="port" value="/dev/ttyUSB0"/>
		<param name="baud" value="57600"/>
	</node>	   
  -->
    <!-- <node name="launchpad_node" pkg="diff_drive" type="launchpad_node.py"/> -->
    
    <node name="lpid_velocity" pkg="diff_drive" type="pid_velocity.py" output="screen">
        <remap from="wheel" to="lwheel"/>
        <remap from="wheel_vtarget" to="single_joint_actuator/lwheel/command"/>
        <remap from="wheel_vel" to="lwheel_vel"/>
        <remap from="motor_cmd" to="vl"/>
        <rosparam param="out_min">-253</rosparam>
        <rosparam param="out_max">253</rosparam>
        <rosparam param="rate">30</rosparam>
        <rosparam param="timeout_ticks">4</rosparam>
        <rosparam param="rolling_pts">5</rosparam>
    </node>
    <node name="rpid_velocity" pkg="diff_drive" type="pid_velocity.py" output="screen">
        <remap from="wheel" to="rwheel"/>
        <remap from="wheel_vtarget" to="single_joint_actuator/rwheel/command"/>
        <remap from="wheel_vel" to="rwheel_vel"/>
        <remap from="motor_cmd" to="vr"/>
        <rosparam param="out_min">-253</rosparam>
        <rosparam param="out_max">253</rosparam>
        <rosparam param="rate">30</rosparam>
        <rosparam param="timeout_ticks">4</rosparam>
        <rosparam param="rolling_pts">5</rosparam>
    </node>
    <node name="joint_vel_service" pkg="diff_drive" type="joints_receive_from_pid_vel.py" output="screen"/>
    <node name="twst_to_motors" pkg="diff_drive" type="twist_to_motors.py" output="screen">
        <remap from="rwheel_vtarget" to="/single_joint_actuator/rwheel/command"/>
        <remap from="lwheel_vtarget" to="/single_joint_actuator/lwheel/command"/>
    </node>
    <node name="robot_hardware_interface" pkg="diff_drive" type="mbot" output="screen"/>
    <node name="diff_tf" pkg="diff_drive" type="diff_tf.py" output="screen">
        <rosparam param="base_frame_id">base_link</rosparam>
        <rosparam param="odom_frame_id">odom</rosparam>
    </node>
    <group>
        <rosparam file="$(find diff_drive)/config/controllers.yaml" command="load"/>
        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"
		  args="
        single_joint_actuator/joint_state_controller
        single_joint_actuator/rwheel
			  single_joint_actuator/lwheel
        "/>
    </group>
</launch>
<!--
   note that we cannot run position controller and velocity controller at the same time, so we create two launch files.
    roslaunch diff_drive check_velocity_controller.launch
    rosrun rosserial_python serial_node.py _port:=/dev/ttyUSB0
    rostopic pub /single_joint_actuator/joint1_velocity_controller/command std_msgs/Float64 "data: 0.57"
    rosrun rqt_reconfigure rqt_reconfigure
    rosrun rqt_plot rqt_plot

  rostopic pub /single_joint_actuator/joint1_velocity_controller/command std_msgs/Float64 "data: 0.6"
-->